### DISCUSSION TITLE
post 요청시 400 에러가 발생합니다.

### DISCUSSION THREAD

**Discussion by kimsw61** [naver/ngrinder#855] [Q&A] [answered]
안녕하세요. jmeter에서 테스트 하던 것을 ngrinder에서 옮겨서 테스트를 진행중에 있습니다.
버전은 3.9.1 이고 다음과 같은 헤더값을 삽입후 테스트를 진행하던 도중 계속 400 에러가 나오고 있습니다.
nodeJS express 프레임워크로 구성된 서버에 https로 전송합니다.

	public void test(){	
		String email = "user@aaa.com";
		String password = "test_password"
		JSONObject testJson= new JSONObject();
		testJson.put("email",email);
		testJson.put("password",password);		
		String jsonDataStr = testJson.toString();

		HTTPResponse result = null;
		try {
                         // 이메일, password 값 post 에 담을 준비
			NVPair param1 = new NVPair("email",email);
			NVPair param2 = new NVPair("password",password);
			NVPair[] params = [ param1, param2 ]

			grinder.logger.info(params.toString())
			result = request.POST(serverUrl, params)
			
			if ( result != null ) {
				if (result.statusCode == 301 || result.statusCode == 302) {
					grinder.logger.warn("Warning. The response may not be correct. The response code was {}.", result.statusCode); 
				} else if (result.statusCode == 400) {
					grinder.logger.error("[RSLT] (400 ERROR) - Request URL : \n" + serverUrl);
					grinder.logger.error("result is " + result.toString());
				} else {
					assertThat(result.statusCode, is(200));
				}
			} else {
				grinder.logger.info("I don't know why result is null" _ jsonDataStr)
			}
		}catch(Exception e) {
			grinder.logger.error("[TRY] Error occured : " + e.toString());
			fail();
		}
	}

헤더값도 체크해보았으나, post body값이 정상적으로 들어가지 않는 것으로 의심됩니다.

어떻게 해야 정상적으로 값을 보낼 수 있을까요?

**Comment by imbyungjun** [comment, marked as answer]
제일 간단한 방법을 소개드리면, `스크립트 - 스크립트 만들기 - 상세 설정 보기 - 헤더`에서 원하는 필드와 값을 추가한 뒤 스크립트를 만들면, 헤더값을 포함하는 테스트 스크립트 템플릿으로 생성됩니다. 
![image](https://user-images.githubusercontent.com/10591327/145767599-f37927de-1173-4456-a722-2b8f35ea847f.png)

**Comment by kimsw61** [comment]
자동 생성된 스크립트에선 {email:\"test@aaa.com\",password:\"my_password\"} 이런 형태로 표시됩니다만, 실제 서버에선 json parse 에러가 나옴을 확인했습니다.
{\"email\":\"test@aaa.com\",\"password\":\"my_password\"} 

이런 식으로 변경해야 되서 나중에 이런식으로 변경이 되었으면 좋을거 같습니다.

**Reply by imbyungjun** [reply]
아 제가 헤더를 세팅하시려는것으로 착각했네요. 
body를 세팅하시려면 `request.POST("http://please_modify_this.com", "{\"hello\":\"world!\"}".getBytes())` 와 같이 수행하시면 됩니다.
